### PR TITLE
Steem5 resourcespatialmanager

### PR DESCRIPTION
**Pull request by htran84**
Add Resource Spatial Manager to atcore

### CODE REVIEW

File: tethysext/atcore/services/resource_spatial_manager.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    def create_extent_layer(self, datastore_name, resource_id):
+        """
+        Creates a GeoServer SQLView Layer for the extent from the resource.
+
+        Args:
+            datastore_name(str): name of the app database, for example: app_primary_db.
+            resource_id(str): id of the extent record.
+        """
+        # Get Default Style Name
+        default_style = f'atcore:{self.VL_EXTENT_TEMPLATE}'
```
Comment (swainn) [inline tethysext/atcore/services/resource_spatial_manager.py:67]: Change `VL_EXTENT_STYLE` definition on line 30 to `VL_EXTENT_STYLE = 'atcore:resource_extent_layer_style'` and change this line to `default_style = self.VL_EXTENT_STYLE`

File: tethysext/atcore/services/resource_spatial_manager.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+            template = Template(text)
+            sql = ' '.join(template.render(sql_context).split())
+
+        # Create SQL View
+        self.gs_api.create_layer(
+            workspace=self.WORKSPACE,
+            datastore_name=datastore_name,
+            feature_name=feature_name,
+            geometry_type=self.GT_POLYGON,
+            srid=4326,
```
Comment (swainn) [inline tethysext/atcore/services/resource_spatial_manager.py:89]: Let's not tie our hands on the SRID. Add `srid=4326` to the arguments of `create_extent_layer` so we have control over it if we decide it needs to be something different in the future.

File: tethysext/atcore/services/resource_spatial_manager.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+            text = sql_template_file.read()
+            template = Template(text)
+            sql = ' '.join(template.render(sql_context).split())
+
+        # Create SQL View
+        self.gs_api.create_layer(
+            workspace=self.WORKSPACE,
+            datastore_name=datastore_name,
+            feature_name=feature_name,
+            geometry_type=self.GT_POLYGON,
```
Comment (swainn) [inline tethysext/atcore/services/resource_spatial_manager.py:88]: Let's add `geometry_type=None` to the arguments of `create_extent_layer` then set it to `self.GT_POLYGON` if it is `None`. Then we'll be able to change it in the future if need be.

File: tethysext/atcore/services/resource_spatial_manager.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            srid=4326,
+            sql=sql,
+            default_style=default_style,
+        )
+
+    def delete_extent_layer(self, datastore_name, resource_id, recurse=True):
+        # feature name
```
Comment (swainn) [inline tethysext/atcore/services/resource_spatial_manager.py:95]: Add doc strings to `delete_extent_layer`. Should be the same as `create_extent_layer`.

File: tethysext/atcore/services/resource_spatial_manager.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        """
+        Get the extent. This will return the list of the extent in EPSG 4326.
+        The query in resource_extent_layer_view transforms all features to 4326.
+
+        Args:
+            datastore_name(str): name of the app database, for example: app_primary_db.
+            resource_id(str): id of the extent record.
+        """
+        # feature name
+        feature_name = f'app_users_resources_{resource_id}'
```
Comment (swainn) [inline tethysext/atcore/services/resource_spatial_manager.py:42]: `f'app_users_resources_{resource_id}'` is used in a few places. Add this function to the class and use it to get the name of the extent layer:

```python
def get_extent_layer_name(self, resource_id):
    """
    Get name of the extent layer for a given resource.

    Args:
        resource_id(str): id of the Resource.

    Returns:
        str: name of the extent layer.

    """
    return f'app_users_resources_extent_{resource_id}'
```

This will ensure we don't end up adding a bug if we change this in the future. Also, not the addition of "extent" to the layer name.

Comment (swainn) [reply tethysext/atcore/services/resource_spatial_manager.py:42]: Use this function everywhere we are hardcoding: `f'app_users_resources_{resource_id}'`

File: tethysext/atcore/services/resource_spatial_manager.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+    def get_projection_string(self, model_db, srid, proj_format=''):
+        pass
+
+    def create_extent_layer(self, datastore_name, resource_id):
+        """
+        Creates a GeoServer SQLView Layer for the extent from the resource.
+
+        Args:
+            datastore_name(str): name of the app database, for example: app_primary_db.
```
[the remark this answers is not in the document]
Comment (swainn) [reply tethysext/atcore/services/resource_spatial_manager.py:63]: Update all `datastore_name` doc strings to match.

File: tethysext/atcore/services/resource_spatial_manager.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+            resource_id(str): id of the extent record.
```
[the remark this answers is not in the document]
Comment (swainn) [reply tethysext/atcore/services/resource_spatial_manager.py:64]: Update all `resource_id` doc strings to match.

File: tethysext/atcore/services/resource_spatial_manager.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+            feature_name=feature_name,
+        )
+
+        return extent
+
+    def get_projection_units(self, resource, srid):
+        pass
+
+    def get_projection_string(self, model_db, srid, proj_format=''):
+        pass
```
Comment (swainn) [inline tethysext/atcore/services/resource_spatial_manager.py:56]: Remove these methods. It will be up to subclasses to implement these.

File: tethysext/atcore/tests/unit_tests/services/resource_spatial_manager.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    @mock.patch('tethysext.atcore.services.base_spatial_manager.GeoServerAPI')
+    def test_create_extent_layer(self, _):
+        sql = "SELECT id, name, description, ST_Transform(extent, 4326) as geometry FROM app_users_resources" \
+              " WHERE id::text = 'test_resource_id'"
+        spatial_manager = ResourceSpatialManager(self.geoserver_engine)
+        spatial_manager.create_extent_layer(datastore_name=self.datastore_name, resource_id=self.resource_id)
+        spatial_manager.gs_api.create_layer.\
+            assert_called_with(datastore_name='test_store', default_style='atcore:resource_extent_layer_style',
+                               feature_name='app_users_resources_test_resource_id', geometry_type='Polygon',
+                               sql=sql, srid=4326, workspace='my-app')
```
Comment (swainn) [inline tethysext/atcore/tests/unit_tests/services/resource_spatial_manager.py:33]: Test for `delete_extent_layer`?

**Review by swainn — CHANGES_REQUESTED** [review]
- [ ] Shouldn't there be changes to the `atcore init` command to create the new vector layer style.
- [ ] Address other comments as well.

File: tethysext/atcore/resources/sld_templates/resource_extent_layer_style.sld [hunk 10]
```diff
@@ -0,0 +1,81 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<sld:StyledLayerDescriptor
+	xmlns="http://www.opengis.net/sld"
+	xmlns:sld="http://www.opengis.net/sld"
+	xmlns:ogc="http://www.opengis.net/ogc"
```
[the remark this answers is not in the document]
Comment (htran84) [reply tethysext/atcore/resources/sld_templates/resource_extent_layer_style.sld:5]: Done

File: tethysext/atcore/resources/sld_templates/resource_extent_layer_style.sld [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+                   <ogc:PropertyIsEqualTo>
+                      <ogc:Function name="geometryType">
+                         <ogc:PropertyName>geometry</ogc:PropertyName>
+                      </ogc:Function>
+                      <ogc:Literal>Polygon</ogc:Literal>
+                   </ogc:PropertyIsEqualTo>
+                   </ogc:Filter>
+					<sld:PolygonSymbolizer>
+						<sld:Stroke>
+							<sld:CssParameter name="stroke">#FFD400</sld:CssParameter>
```
[the remark this answers is not in the document]
Comment (htran84) [reply tethysext/atcore/resources/sld_templates/resource_extent_layer_style.sld:53]: Done

File: tethysext/atcore/services/resource_spatial_manager.py [hunk 12]
```diff
@@ -0,0 +1,103 @@
+"""
+********************************************************************************
+* Name: model_db_spatial_manager
+* Author: nswain
+* Created On: July 06, 2018
+* Updated on: December 19, 2018
+* Copyright: (c) Aquaveo 2018
```
[the remark this answers is not in the document]
Comment (htran84) [reply tethysext/atcore/services/resource_spatial_manager.py:7]: Done

File: tethysext/atcore/services/resource_spatial_manager.py [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+********************************************************************************
+"""
+import os
+from jinja2 import Template
+
+from abc import abstractmethod
```
[the remark this answers is not in the document]
Comment (htran84) [reply tethysext/atcore/services/resource_spatial_manager.py:13]: Done

File: tethysext/atcore/services/resource_spatial_manager.py [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+
+from tethysext.atcore.services.exceptions import UnitsNotFound, UnknownUnits
+from tethysext.atcore.services.base_spatial_manager import BaseSpatialManager
+from tethysext.atcore.models.app_users import Resource
+
+
+class ResourceSpatialManager(BaseSpatialManager):
+    """
+    Class for Spatial Managers using a postgres database.
```
[the remark this answers is not in the document]
Comment (htran84) [reply tethysext/atcore/services/resource_spatial_manager.py:22]: Done

File: tethysext/atcore/services/resource_spatial_manager.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (htran84) [reply tethysext/atcore/services/resource_spatial_manager.py:64]: Done

File: tethysext/atcore/services/resource_spatial_manager.py (hunk 1 above)
Comment (htran84) [reply tethysext/atcore/services/resource_spatial_manager.py:67]: Done

File: tethysext/atcore/services/resource_spatial_manager.py (hunk 4 above)
Comment (htran84) [reply tethysext/atcore/services/resource_spatial_manager.py:95]: Done

File: tethysext/atcore/services/resource_spatial_manager.py (hunk 8 above)
Comment (htran84) [reply tethysext/atcore/services/resource_spatial_manager.py:56]: Done

File: tethysext/atcore/tests/unit_tests/services/resource_spatial_manager.py (hunk 9 above)
Comment (htran84) [reply tethysext/atcore/tests/unit_tests/services/resource_spatial_manager.py:33]: Add test for delete_extent_layer

File: tethysext/atcore/services/resource_spatial_manager.py [hunk 15]
```diff
@@ -49,25 +58,24 @@ def get_extent_for_project(self, datastore_name, resource_id):
...
         Creates a GeoServer SQLView Layer for the extent from the resource.
 
         Args:
             datastore_name(str): name of the app database, for example: app_primary_db.
-            resource_id(str): id of the extent record.
+            resource_id(str): id of the Resources.
+            geometry_type(str): type of geometry. Pick from: Polygon, LineString, Point.
         """
+        geometry_check_list = [x.lower() for x in [self.GT_POLYGON, self.GT_LINE, self.GT_POINT]]
+        if geometry_type is None or geometry_type.lower() not in geometry_check_list:
```
Comment (swainn) [inline tethysext/atcore/services/resource_spatial_manager.py:71]: Default to GT_POLYGON if no value is given. Raise a ValueError if an invalid value is given so the behavior isn't confusing.

File: tethysext/atcore/services/resource_spatial_manager.py (hunk 5 above)
Comment (swainn) [reply tethysext/atcore/services/resource_spatial_manager.py]: Please change 'app_users_resources_' to 'app_users_resources_extent_'. This will future proof the layer name if we end up having other layers from the app_users_resources table.

File: tethysext/atcore/services/resource_spatial_manager.py [hunk 16]
```diff
@@ -1,33 +1,42 @@
...
+        Get name of the extent layer for a given resource.
+
+        Args:
+            resource_id(str): id of the Resource.
+
+        Returns:
+            str: name of the extent layer.
+
+        """
+        return f'app_users_resources_{resource_id}'
```
Comment (swainn) [inline tethysext/atcore/services/resource_spatial_manager.py:39]: Please change 'app_users_resources_' to 'app_users_resources_extent_'. This will future proof the layer name if we end up having other layers from the app_users_resources table.

File: tethysext/atcore/services/resource_spatial_manager.py [hunk 17]
```diff
@@ -58,19 +59,25 @@ def get_extent_for_project(self, datastore_name, resource_id):
...
-            datastore_name(str): name of the app database, for example: app_primary_db.
+            datastore_name(str): name of the PostGIS datastore in GeoServer that contains the layer data.
+             For example: app_primary_db.
             resource_id(str): id of the Resources.
             geometry_type(str): type of geometry. Pick from: Polygon, LineString, Point.
+            srid(str): Spatial Reference Identifier of the extent. Default to 4326.
         """
         geometry_check_list = [x.lower() for x in [self.GT_POLYGON, self.GT_LINE, self.GT_POINT]]
-        if geometry_type is None or geometry_type.lower() not in geometry_check_list:
+        if geometry_type is None:
```
Comment (swainn) [inline tethysext/atcore/services/resource_spatial_manager.py:74]: Need to add a test for this case.

File: tethysext/atcore/services/resource_spatial_manager.py [hunk 18]
```diff
@@ -58,19 +59,25 @@ def get_extent_for_project(self, datastore_name, resource_id):
...
             geometry_type = self.GT_POLYGON
 
+        if geometry_type.lower() not in geometry_check_list:
```
Comment (swainn) [inline tethysext/atcore/services/resource_spatial_manager.py:77]: Need to add a test for this case.

File: tethysext/atcore/services/resource_spatial_manager.py [hunk 19]
```diff
@@ -79,6 +86,7 @@ def create_extent_layer(self, datastore_name, resource_id, geometry_type=None):
 
         sql_context = {
             'resource_id': resource_id,
+            'srid': srid,
```
Comment (swainn) [inline tethysext/atcore/services/resource_spatial_manager.py]: There are two SRID's we are dealing with here. The SRID in the SQL is the SRID we'd like the geometry to be transformed to when it is queried. The SRID we specify in the create_layer call is the SRID of the given features. We don't want to lock these to be the same SRID. I'd say leave the SRID in the SQL hardcoded to transform to 4326 when we get the layer. But allow the user to specify the SRID of the features they have provided, defaulting to 4326. Make sense?

TL;DR: Don't pass the SRID argument to the SQL template.

**Review by swainn — CHANGES_REQUESTED** [review]

File: tethysext/atcore/services/resource_spatial_manager.py [hunk 20]
```diff
@@ -77,35 +77,35 @@ def create_extent_layer(self, datastore_name, resource_id, geometry_type=None, s
...
-        self.gs_api.create_layer(
-            workspace=self.WORKSPACE,
-            datastore_name=datastore_name,
-            feature_name=feature_name,
-            geometry_type=self.GT_POLYGON,
-            srid=srid,
-            sql=sql,
-            default_style=default_style,
-        )
+        else:
```
Comment (swainn) [inline tethysext/atcore/services/resource_spatial_manager.py:80]: This `else` is unecessary. The `raise ValueError` will prevent it from going on when needed.

File: tethysext/atcore/services/resource_spatial_manager.py (hunk 19 above)
Comment (swainn) [reply tethysext/atcore/services/resource_spatial_manager.py:89]: I was wrong on this. Sorry for the confusion.

**Review by swainn — APPROVED** [review]
